prow/crier/controller.go

```diff
@@ -0,0 +1,183 @@
+	if !cache.WaitForCacheSync(stopCh, c.HasSynced) {
+		utilruntime.HandleError(fmt.Errorf("Error syncing cache"))
+		return
+	}
+	logrus.Info("Controller.Run: cache sync complete")
+
+	// run the runWorker method every second with a stop channel
+	wait.Until(c.runWorker, time.Second, stopCh)
+}
+
+// HasSynced allows us to satisfy the Controller interface
+// by wiring up the informer's HasSynced method to it
+func (c *Controller) HasSynced() bool {
+	return c.informer.Informer().HasSynced()
+}
+
+// runWorker executes the loop to process new items added to the queue
+func (c *Controller) runWorker() {
+	logrus.Info("Controller.runWorker: starting")
+
+	// invoke processNextItem to fetch and consume the next change
+	// to a watched or listed resource
+	for c.processNextItem() {
+		logrus.Info("Controller.runWorker: processing next item")
+	}
+
+	logrus.Info("Controller.runWorker: completed")
+}
+
+// processNextItem retrieves each queued item and takes the
+// necessary handler action based off of if the item was
+// created or deleted
+func (c *Controller) processNextItem() bool {
+	// fetch the next item (blocking) from the queue to process or
+	// if a shutdown is requested then return out of this to stop
+	// processing
+	key, quit := c.queue.Get()
+
+	// stop the worker loop from running as this indicates we
+	// have sent a shutdown message that the queue has indicated
+	// from the Get method
+	if quit {
+		return false
+	}
+
+	defer c.queue.Done(key)
+
+	// assert the string out of the key (format `namespace/name`)
+	keyRaw := key.(string)
+	namespace, name, err := cache.SplitMetaNamespaceKey(keyRaw)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", keyRaw))
+	}
+
+	// take the string key and get the object out of the indexer
+	//
+	// item will contain the complex object for the resource and
+	// exists is a bool that'll indicate whether or not the
+	// resource was created (true) or deleted (false)
+	//
+	// if there is an error in getting the key from the index
+	// then we want to retry this particular queue key a certain
+	// number of times (5 here) before we forget the queue key
+	// and throw an error
+	pj, err := c.informer.Lister().ProwJobs(namespace).Get(name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logrus.Infof("Controller.processNextItem: object no longer exist: %s", keyRaw)
+			c.queue.Forget(key)
+			return true
+		}
+
+		if c.queue.NumRequeues(key) < 5 {
+			logrus.Errorf("Controller.processNextItem: Failed processing item with key %s with error %v, retrying", key, err)
+			c.queue.AddRateLimited(key)
+		} else {
+			logrus.Errorf("Controller.processNextItem: Failed processing item with key %s with error %v, no more retries", key, err)
+			c.queue.Forget(key)
+			utilruntime.HandleError(err)
+		}
+		return true
+	}
+
+	logrus.Infof("Controller.processNextItem: object created & updated detected: %s", pj.Spec.Job)
```

Comment:
Technically just `WithFields(fields(pj)).Info("reconciling ProwJob")`? Why we're processing is an implementation detail